### PR TITLE
[TEP-0133]: Configure Default Resolver

### PR DESCRIPTION
**Pull request by QuanZhang-William**
# Changes

This commit introduces a new `default-resolver-type` field to the `config-defaults` ConfigMap, which configures the default resolver type to be used when the `resolver` is not explicitly provided in the input. Supporting the default resolver type improves simplicity at the authoring time. More details can be found in [TEP-0113: Configure Default Resolver].

/kind feature

[TEP-0113: Configure Default Resolver]: https://github.com/tektoncd/community/blob/main/teps/0133-configure-default-resolver.md
# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [x] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
[TEP-0133] Add "default-resolver-type" field in the "default-configs" ConfigMap to configure default resolver
```

### CODE REVIEW

**Comment by QuanZhang-William** [comment]
Tracking issue: https://github.com/tektoncd/pipeline/issues/5907

**Review by lbernick — COMMENTED** [review]
thanks @QuanZhang-William! Is there also validation for taskrefs with a default resolver configured?

**Comment by QuanZhang-William** [comment]
> thanks @QuanZhang-William! Is there also validation for taskrefs with a default resolver configured?

This is a good question and it is a bit tricky. The short answer is no, and I feel it is not necessary?:

1) If the default resolver is ***NOT*** configured, we already have validations and test cases cover it.
2) If the default resolver is configured, the value will be populated at *Mutation Webhook* phase which happens before the validation. So the validation will not receive a request without the resolver value when the default is set (which I think is the scenario you talked about to validation).

File: docs/resolution.md [hunk 1]
```diff
@@ -21,6 +21,8 @@ For new users getting started with Tekton Pipeline remote resolution, check out
 These resolvers are enabled by setting the appropriate feature flag in the `resolvers-feature-flags`
 ConfigMap in the `tekton-pipelines-resolvers` namespace. See the [section in install.md](install.md#configuring-built-in-remote-task-and-pipeline-resolution) for details.
 
+The default resolver type can be configured by the `default-resolver-type` field in the `config-defaults` ConfigMap. See [additional-configs.md](./additional-configs.md) for details.
```
[the remark this answers is not in the document]
Comment (lbernick) [reply docs/resolution.md:24]: One example is trusted resources, which is guarded by the flag `resource-verification-mode`. You would just need to add it to the list of alpha features you linked, and describe it as an alpha feature in the docs.

**Comment by lbernick** [comment]
> > thanks @QuanZhang-William! Is there also validation for taskrefs with a default resolver configured?
> 
> This is a good question and it is a bit tricky. The short answer is no, and I feel it is not necessary?:
> 
> 1. If the default resolver is _**NOT**_ configured, we already have validations and test cases cover the empty resolver scenario.
> 2. If the default resolver is configured, the value will be populated at _Mutation Webhook_ phase which happens before the validation, So the validation will never receive a request which empty resolver in this case (which I think is the scenario you talked about to validation).

ah that's a good point-- I wish there was a better way to unit test this, but this seems fine.

just a quick comment, can you add the TEP in the release note?

File: pkg/apis/pipeline/v1/pipeline_defaults_test.go [hunk 2]
```diff
@@ -120,6 +125,72 @@ func TestPipelineSpec_SetDefaults(t *testing.T) {
...
+		wc: func(ctx context.Context) context.Context {
+			s := config.NewStore(logtesting.TestLogger(t))
+			s.OnConfigChanged(&corev1.ConfigMap{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: config.GetDefaultsConfigName(),
+				},
+				Data: map[string]string{
+					"default-resolver-type": "git",
+				},
+			})
```
[the remark this answers is not in the document]
Comment (QuanZhang-William) [reply pkg/apis/pipeline/v1/pipeline_defaults_test.go]: Thanks, did some refactoring based on your comment, PTAL!

File: docs/resolution.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (QuanZhang-William) [reply docs/resolution.md]: Adde as a alpha feature flag!

File: pkg/apis/config/testing/defaults.go [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+	"testing"
+
+	"github.com/tektoncd/pipeline/pkg/apis/config"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	logtesting "knative.dev/pkg/logging/testing"
+)
+
+// SetDefaults sets the default ConfigMap values in an existing context (for use in testing)
+func SetDefaults(ctx context.Context, t *testing.T, data map[string]string) context.Context {
```
Comment (lbernick) [inline pkg/apis/config/testing/defaults.go:14]: thanks for adding the helper function! would you mind moving the refactoring into a separate PR? other than that this LGTM

**Review by chuangw6 — COMMENTED** [review]
Thanks @QuanZhang-William . It seems like that setting default resolver for `PipelineRef` in a PipelineRun is missing.

**Comment by chuangw6** [comment]
> Thanks @QuanZhang-William . It seems like that setting default resolver for `PipelineRef` in a PipelineRun is missing.

With that, we should also consider the special case that the pipeline spec referred by `PipelineRef` also references other remote tasks. Example:

```yaml
PipelineRun
...
spec:
   pipelineRef:
       resolver: git # if missing, the default resolver can be set by the mutating webhook
       params:
            ....
...
```

```yaml
Pipeline referenced by the above PipelineRun
...
spec:
     tasks:
        - name: clone-source
          taskRef:
               resolver: bundles # how to set default resolver for this?
               params:
                   ...
```

Setting default resolver for `PipelineRef` can be done in mutating webhook, but how do we set default resolver for its inner `TaskRef` used by the remote pipeline? 
It seems like it's already handled by the reconciler by [recalling](https://github.com/tektoncd/pipeline/blob/358501b4317e44f860aef5d323dbbcaaf338cff2/pkg/reconciler/pipelinerun/pipelinespec/pipelinespec.go#L70) `pipelineSpec.SetDefaults(ctx)` after fetching the remote pipeline spec. 

If this worked as intended, I think it's worth adding test cases for it and mentioning this special case in the TEP. wdyt?

File: pkg/apis/config/testing/defaults.go (hunk 3 above)
Comment (QuanZhang-William) [reply pkg/apis/config/testing/defaults.go:14]: Sure, created a separate PR and updates the usage of the new helper function PTAL!: https://github.com/tektoncd/pipeline/pull/6339

/hold unitil https://github.com/tektoncd/pipeline/pull/6339 is merged

File: pkg/reconciler/pipelinerun/pipelinespec/pipelinespec_test.go [hunk 4]
```diff
@@ -127,65 +128,116 @@ func TestGetPipelineSpec_Invalid(t *testing.T) {
...
-				Name: "tref",
+		{
+			name:               "resolve remote task with taskRef resolver - default resolver configured",
+			sourceMeta:         sourceMeta,
+			sourceConfigSource: sourceConfigSource,
+			pr: &v1beta1.PipelineRun{
+				Spec: v1beta1.PipelineRunSpec{
+					PipelineRef: &v1beta1.PipelineRef{
+						ResolverRef: v1beta1.ResolverRef{
+							Resolver: "foo",
```
Comment (QuanZhang-William) [inline pkg/reconciler/pipelinerun/pipelinespec/pipelinespec_test.go:189]: The `PipelineRun` resolver must have been pre-populated at this point

**Comment by QuanZhang-William** [comment]
> > Thanks @QuanZhang-William . It seems like that setting default resolver for `PipelineRef` in a PipelineRun is missing.
> 
> With that, we should also consider the special case that the pipeline spec referred by `PipelineRef` also references other remote tasks. Example:
> 
> ```yaml
> PipelineRun
> ...
> spec:
>    pipelineRef:
>        resolver: git # if missing, the default resolver can be set by the mutating webhook
>        params:
>             ....
> ...
> ```
> 
> ```yaml
> Pipeline referenced by the above PipelineRun
> ...
> spec:
>      tasks:
>         - name: clone-source
>           taskRef:
>                resolver: bundles # how to set default resolver for this?
>                params:
>                    ...
> ```
> 
> Setting default resolver for `PipelineRef` can be done in mutating webhook, but how do we set default resolver for its inner `TaskRef` used by the remote pipeline? It seems like it's already handled by the reconciler by [recalling](https://github.com/tektoncd/pipeline/blob/358501b4317e44f860aef5d323dbbcaaf338cff2/pkg/reconciler/pipelinerun/pipelinespec/pipelinespec.go#L70) `pipelineSpec.SetDefaults(ctx)` after fetching the remote pipeline spec.
> 
> _**If**_ this worked as intended, I think it's worth adding test cases for it and mentioning this special case in the TEP. wdyt?

Thanks @chuangw6! Good catch on the missing `PipelineRef` change! I have added it and PTAL.

Regarding the edge case, yes, this is intended and I have added a test case for it. I think the [examples given in the tep](https://github.com/tektoncd/community/blob/main/teps/0133-configure-default-resolver.md#proposal) has elaborated this idea.

File: pkg/apis/config/testing/defaults.go (hunk 3 above)
Comment (QuanZhang-William) [reply pkg/apis/config/testing/defaults.go]: https://github.com/tektoncd/pipeline/pull/6339 is merged

/hold cancel

**Comment by Yongxuanzhang** [comment]
I wonder if we're missing unit tests in `pkg/apis/config/default_test.go` ?  It reads config from file and compare if it is expected. 
Other code lgtm

**Comment by QuanZhang-William** [comment]
> I wonder if we're missing unit tests in `pkg/apis/config/default_test.go` ? It reads config from file and compare if it is expected. Other code lgtm

Thanks @Yongxuanzhang, just added the tests, PTAL!

**Comment by Yongxuanzhang** [comment]
Thank you!  Test coverage also improved!
/lgtm

**Comment by dibyom** [comment]
/approve
/hold 

(Adding a hold in case another approver wants to review this PR but I think this is good to go)

**Comment by pritidesai** [comment]
(FYI: I am working with a limited internet/cellular connectivity in the Bay Area)

how does this impact existing tasks and pipelines? The default is set to ‘git’ but how about params? How does this add value when I have to specify params for git? Please note that I am not objecting to this change, but trying to understand the impact of this.

**Comment by QuanZhang-William** [comment]
> (FYI: I am working with a limited internet/cellular connectivity in the Bay Area)
> 
> how does this impact existing tasks and pipelines? The default is set to ‘git’ but how about params? How does this add value when I have to specify params for git? Please note that I am not objecting to this change, but trying to understand the impact of this.

Hi @pritidesai, there is no impact to existing `Task` and `Pipeline` as all the resources today have the `resolver` field specified explicitly.

Simplicity is the main benefit of this feature. This feature can be used together with the default parameters we support for resolvers, for example: 

by configuring the default values like

``` yaml
  # Setting this flag to a resolver type as the default resolver
  default-resolver-type: "git"
  
  # The below fields are git resolver specific and have been supported already
  # The git url to fetch the remote resource from when using anonymous cloning.
  default-url: "https://github.com/tektoncd/catalog.git"
  # The git revision to fetch the remote resource from with either anonymous cloning or the authenticated API.
  default-revision: "main"
```

the author can reduce a task run from 
```
apiVersion: tekton.dev/v1beta1
kind: TaskRun
metadata:
 name: demo-task-run
spec:
 taskRef:
   resolver: git
   params:
   - name: url
     value: https://github.com/tektoncd/catalog.git
   - name: revision
     value: main
   - name: pathInRepo
     value: task/golang-build/0.1/golang-build.yaml
...
```

to 
```
apiVersion: tekton.dev/v1beta1
kind: TaskRun
metadata:
 name: demo-task-run
spec:
 taskRef:
  params:
  - name: pathInRepo
    value: task/golang-build/0.1/golang-build.yaml
...
```

The use cases are discuss in [TEP-0133](https://github.com/tektoncd/community/blob/main/teps/0133-configure-default-resolver.md#use-cases)

**Comment by dibyom** [comment]
/hold cancel

Chatted with Priti earlier in the day, I think we can merge once it has been rebased
